src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -2801,6 +2803,86 @@ exprt c_typecheck_baset::do_special_functions(
     overflow.type() = bool_typet{};
     return overflow;
   }
+  else if(
+    identifier == "__builtin_add_overflow" ||
+    identifier == "__builtin_sub_overflow" ||
+    identifier == "__builtin_mul_overflow")
+  {
+    if(expr.arguments().size() != 3)
+    {
+      std::ostringstream error_message;
+      error_message << expr.source_location().as_string() << ": " << identifier
+                    << " takes exactly 3 arguments, but "
+                    << expr.arguments().size() << " were provided";
+      throw invalid_source_file_exceptiont{error_message.str()};
+    }
+
+    for(std::size_t arg_index = 0; arg_index < 3; ++arg_index)
+    {
+      auto const &argument = expr.arguments()[arg_index];
+      bool const arg_has_correct_type =
+        (arg_index < 2) ? (argument.type().id() == ID_signedbv ||
+                           argument.type().id() == ID_unsignedbv)
+                        : (argument.type().id() == ID_pointer &&
+                           (argument.type().subtype().id() == ID_signedbv ||
+                            argument.type().subtype().id() == ID_unsignedbv));
+      if(!arg_has_correct_type)
+      {
+        std::ostringstream error_message;
+        error_message << expr.source_location().as_string() << ": "
+                      << identifier << " has signature " << identifier
+                      << "(integral, integral, integral*), "
+                      << "but argument " << (arg_index + 1) << " ("
+                      << expr2c(argument, *this) << ") has type `"
+                      << type2c(argument.type(), *this) << '`';
+        throw invalid_source_file_exceptiont{error_message.str()};
+      }
+    }
+
+    auto lhs = expr.arguments()[0];
+    auto rhs = expr.arguments()[1];
+    auto result_ptr = expr.arguments()[2];
+
+    typecheck_expr(lhs);
+    typecheck_expr(rhs);
+    typecheck_expr(result_ptr);
+
+    auto const make_operation = [&identifier](exprt lhs, exprt rhs) -> exprt {
+      if(identifier == "__builtin_add_overflow")
+      {
+        return plus_exprt{lhs, rhs};
+      }
+      else if(identifier == "__builtin_sub_overflow")
+      {
+        return minus_exprt{lhs, rhs};
+      }
+      else
+      {
+        INVARIANT(
+          identifier == "__builtin_mul_overflow",
```

Comment:
Given this is inside the `else if(
     identifier == "__builtin_add_overflow" ||
     identifier == "__builtin_sub_overflow" ||
     identifier == "__builtin_mul_overflow")` branch, is it any easier to read to just make this final else and 'else if(identifier == blah)` and then and a final `else UNREACHABLE` ?

Reply:
I don’t know about “easier”, the reason I wrote it like this is because I don’t like seeing “unreachable” whenever one of these changes. So I have a slight preference for doing it this way round, but I don’t mind doing it the other way round either.

Reply:
Fair enough, I don't have strong leaning either way.